R/stacking.R

```diff
@@ -0,0 +1,1626 @@
+
+    n.manual.variables <- max(vapply(manual.stacking.obs.groups, length,
+                                     integer(1)))
+
+    # Remove rows in new.stacking.groups that contain overlap with
+    # manual.stacking.groups
+    for (manual.obs.group in manual.stacking.obs.groups)
+    {
+        manual.obs.group.without.na <- removeNA(manual.obs.group)
+        is.overlapping <- apply(new.stacking.groups, 1, function(obs.group) {
+            any(manual.obs.group.without.na %in% obs.group)
+        })
+        new.stacking.groups <- new.stacking.groups[!is.overlapping, , drop = FALSE]
+    }
+
+    # Increase number of columns in new.stacking.groups if necessary to fit
+    # manual.stacking.groups
+    n.rows <- nrow(new.stacking.groups)
+    n.cols <- ncol(new.stacking.groups)
+    max.stacking <- max(length(manual.stacking.obs.groups), n.cols)
+    m <- matrix(nrow = n.rows, ncol = max.stacking)
+    m[seq_len(n.rows), seq_len(n.cols)] <- new.stacking.groups
+    new.stacking.groups <- m
+
+    # Append manual.stacking.groups to stacking.groups
+    ind <- seq_along(manual.stacking.obs.groups)
+    for (i in seq_len(n.manual.variables))
+    {
+        new.group <- rep(NA_integer_, max.stacking)
+        new.group[ind] <- vapply(manual.stacking.obs.groups, `[`,
+                                 integer(1), i)
+        non.missing.ind <- removeNA(new.group)
+        if (!allIdentical(v.types[non.missing.ind]) ||
+            !allIdentical(v.val.attr[non.missing.ind]))
+        {
+            warning("No manual stacking was conducted as the manual ",
+                    "stacking input '", input.text, "' would result in the ",
+                    "stacking of variables with mismatching types or ",
+                    "value attributes.")
+            return(stacking.groups)
+        }
+
+        new.stacking.groups <- rbind(new.stacking.groups, new.group)
+    }
+
+    is.manually.stacked <- c(rep(FALSE, nrow(new.stacking.groups) - n.manual.variables),
+                             rep(TRUE, n.manual.variables))
+    attr(new.stacking.groups, "is.manually.stacked") <- is.manually.stacked
+
+    new.stacking.groups
+}
+
+# Constructs the stacked data set as a data frame from the input data set and
+# the stacking.groups matrix
+stackedDataSet <- function(input.data.set, input.data.set.metadata,
+                           stacking.groups, include.original.case.variable,
+                           include.observation.variable,
+                           common.labels.list)
+{
+    if (is.null(stacking.groups) || nrow(stacking.groups) == 0)
+    {
+        return(data.frame(lapply(input.data.set, function(v) {
+            attr(v, "is.stacked") <- FALSE
+            attr(v, "is.manually.stacked") <- NA
+            v
+        })))
+    }
+
+    input.v.names <- input.data.set.metadata$variable.names
+    input.v.labels <- input.data.set.metadata$variable.labels
+    retained.indices <- retainedIndices(stacking.groups,
```

Comment:
Indices into what space?  I think it's the original data file.  Your variable names need to be more explicit where they are calculated in a remote function and passed around.

Reply:
Improved names in 57a6712